### PR TITLE
TST: Added iloc tests for casting from object to numeric dtypes for duplicate columns

### LINKED ISSUE #22035: Unable to change dtype of a column after concat and duplicate column names
#### Code Sample, a copy-pastable example if possible

```python
import pandas

a = pandas.DataFrame({'a': ['0'], 'b': ['str']})

print(a.dtypes)

b = pandas.concat([a, pandas.DataFrame({'b': ['str2']})], axis=1)

print(b.dtypes)

b.iloc[:, 0] = [int(v) for v in a.iloc[:, 0]]

print(b.dtypes)
```

#### Problem description

The issue seems to be that if there is a DataFrame with duplicate column names, I cannot modify an (unrelated) column and convert its dtype from object to int.

#### Expected Output

The final print should be:

```
a     int64
b    object
b    object
dtype: object
```

And not:

```
a    object
b    object
b    object
dtype: object
```

It is interesting that if I change concat line to (see renaming of column `b` to `c`):

```
b = pandas.concat([a, pandas.DataFrame({'c': ['str2']})], axis=1)
```

Then the output is correctly:

```
a     int64
b    object
c    object
dtype: object
```

So it seems it has something with the fact that there are duplicate column names on other two columns, not the one being changed.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.13.0-46-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.3
pytest: None
pip: 18.0
setuptools: 40.0.0
Cython: None
numpy: 1.15.0
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by isaac-chung**
- [x] closes #22035
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexing/test_iloc.py [hunk 1]
```diff
@@ -1186,6 +1186,14 @@ def test_iloc_setitem_series_duplicate_columns(self):
         df.iloc[:, 0] = df.iloc[:, 0].astype(np.float64)
         assert df.dtypes.iloc[2] == np.int64
 
+    def test_iloc_setitem_dtypes_duplicate_columns(self):
+        df = DataFrame([["0", "str", "1.2"]], columns=["a", "b", "b"])
+        df.iloc[:, 0] = df.iloc[:, 0].astype("int64")
+        assert df.dtypes.iloc[0] == np.int64
```
Comment (phofl) [inline pandas/tests/indexing/test_iloc.py:1192]: Please parametrize and check df with tm.assert_frame_equal and also add gh reference

Comment (isaac-chung) [reply pandas/tests/indexing/test_iloc.py]: Added tm.assert_frame_equal check and gh reference.

File: pandas/tests/indexing/test_iloc.py [hunk 2]
```diff
@@ -1186,6 +1186,17 @@ def test_iloc_setitem_series_duplicate_columns(self):
         df.iloc[:, 0] = df.iloc[:, 0].astype(np.float64)
         assert df.dtypes.iloc[2] == np.int64
 
+    def test_iloc_setitem_dtypes_duplicate_columns(self):
+        # GH#22035
+        df = DataFrame([["0", "str", "1.2"]], columns=["a", "b", "b"])
```
Comment (phofl) [inline pandas/tests/indexing/test_iloc.py:1191]: please paramterize over dtypes

**Comment by phofl** [comment]
You could install pre-commit to automatically run the linting checks before committing

File: pandas/tests/indexing/test_iloc.py (hunk 2 above)
Comment (isaac-chung) [reply pandas/tests/indexing/test_iloc.py]: My apologies for missing the parametrization earlier

**Comment by isaac-chung** [comment]
I assume an extended test would not fall into the "whatsnew entry", so checking it off for completeness. Is there anything I can do about the failed checks?

**Comment by jreback** [comment]
> I assume an extended test would not fall into the "whatsnew entry", so checking it off for completeness. Is there anything I can do about the failed checks?

no

**Comment by jreback** [comment]
looks fine ping on green.

**Review by phofl — APPROVED** [review]

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @isaac-chung! The CI failure was unrelated

**Pull request merged**